### PR TITLE
Unify MergeTree Walks into Single Interative Function

### PR DESCRIPTION
**Pull request by anthony-murphy**
## Description

In merge tree there were multiple near duplicate methods for walking the tree. Additionally, all these methods were recursive. This change unifies the tree walks into a single method, that is iterative rather than recursive. The primary motivation for this change in related to #1009 where it is necessary to do tree walks to shift local references for each undo/redo, as the tree grew these walks grew deeper, and the repeated stack allocations of the recursive calls lead to performance issues. In addition, removing all the duplicated variations will make maintenance easier, and should reduce bundle size.

### CODE REVIEW

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -837,7 +844,7 @@ export class MergeTree {
             offset = start;
             return false;
         };
-        this.searchBlock(this.root, pos, 0, refSeq, clientId, { leaf }, undefined, localSeq);
+        this.nodeMap(refSeq, clientId, leaf, undefined, undefined, pos, pos + 1, localSeq);
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:847]: search block was the wrong method here, as it stops at every segment, whereas nodemap uses the tree to optimally find the position.

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 2]
```diff
@@ -2283,98 +2214,79 @@ export class MergeTree {
...
-                return true;
-            }
+        refSeq: number,
+        clientId: number,
+        leaf: ISegmentAction<TClientData>,
+        accum: TClientData,
+        post?: BlockAction<TClientData>,
+        start: number = 0,
+        end?: number,
+        localSeq?: number,
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:2224]: bonus. adding local seq here was cheap, which helps get containing segment, which helps for rebase, rollback, re-connect etc

File: packages/dds/merge-tree/src/mergeTreeNodeWalk.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                ? block.children[forward ? 0 : childCount - 1]
+                : undefined;
+        }
+
+        let exit = blockResult === NodeAction.Exit;
+
+        // walk the leaves if we reached them
+        if (start !== undefined) {
+            for (let i = start.index; i !== -1 && i !== childCount; i += increment) {
+                // the above loop ensues start is a leaf or undefined, so all children
```
Comment (scarlettjlee) [inline packages/dds/merge-tree/src/mergeTreeNodeWalk.ts:74]: typo: "ensures"

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTreeNodeWalk.ts:75]:
```suggestion
                // the above loop ensures start is a leaf or undefined, so all children
```

File: .vscode/launch.json [hunk 4]
```diff
@@ -284,7 +284,7 @@
             "name": "Debug Current Mocha Test",
             "env": {
                 // "fluid__test__driver": "odsp" //values: local, tinylicious, routerlicious, odsp,
-                // "FLUID_TEST_VERBOSE": "1"
+                "FLUID_TEST_VERBOSE": "1"
```
Comment (anthony-murphy) [inline .vscode/launch.json:287]:
```suggestion
                // "FLUID_TEST_VERBOSE": "1"
```

File: packages/dds/merge-tree/src/mergeTreeNodeWalk.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    leafAction: (seg: ISegment) => boolean | undefined,
+): boolean {
+    if (startNode.parent === undefined) {
+        return true;
+    }
+
+    return depthFirstNodeWalk(
+        startNode.parent,
+        // this will either be the sibling, or undefined
+        // either if fine, and will result in skipping
```
Comment (scarlettjlee) [inline packages/dds/merge-tree/src/mergeTreeNodeWalk.ts:131]:
```suggestion
        // either is fine, and will result in skipping
```

File: packages/dds/merge-tree/src/mergeTreeNodeWalk.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        // if there is no upAction, we don't need to walk up before exiting
+        if (upAction === undefined && exit) {
+            return false;
+        }
+
+        // since we already enumerated the children
+        // we walk up to the next level until there is a sibling
+        // of all the way up if exit is true
```
Comment (scarlettjlee) [inline packages/dds/merge-tree/src/mergeTreeNodeWalk.ts:91]:
```suggestion
        // or all the way up if exit is true
```

File: packages/dds/merge-tree/src/mergeTreeNodeWalk.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    startNode: IMergeNode,
+    leafAction: (seg: ISegment) => boolean | undefined,
+): boolean {
+    if (startNode.parent === undefined) {
+        return true;
+    }
+    return depthFirstNodeWalk(
+        startNode.parent,
+        // this will either be the sibling, or undefined
+        // either if fine, and will result in skipping
```
Comment (scarlettjlee) [inline packages/dds/merge-tree/src/mergeTreeNodeWalk.ts:153]:
```suggestion
        // either is fine, and will result in skipping
```

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 1 above)
Comment (scarlettjlee) [reply packages/dds/merge-tree/src/mergeTree.ts:848]: What consequences did that have? perf?

**Comment by scarlettjlee** [comment]
Why merge into next rather than main?

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:848]: walking the tree is with node walk is O(log N), whereas walking all segments with search will be O(N), so yeah it was perf

**Comment by anthony-murphy** [comment]
> Why merge into next rather than main?

I'm breaking out smaller pieces of a much larger pr that is based on next. all the smaller prs are layered together so it easiest to not have to rebase any to main and wait for them to come upstream.

**Review by scarlettjlee — APPROVED** [review]

**Pull request merged**